### PR TITLE
fix update_velocity

### PR DESCRIPTION
**Pull request by td1803**
Update für Recycleview / Scrollview not smoothe.

Friction is negative acceleration. So the resulting velocity must be calculated using friction
over time, and not a constant factor. Important here are the big jumps in the factor dt on Samsung S5.
This leads to corresponding irregular behaviour of the view.

This is what i get on Windows:

update_velocity: velocity:6064.52 dt: 0.01
update_velocity: velocity:5761.29 dt: 0.02
update_velocity: velocity:5473.23 dt: 0.02
update_velocity: velocity:5199.57 dt: 0.02
update_velocity: velocity:4939.59 dt: 0.02
update_velocity: velocity:4692.61 dt: 0.02
update_velocity: velocity:4457.98 dt: 0.01
update_velocity: velocity:4235.08 dt: 0.02
update_velocity: velocity:4023.33 dt: 0.02
update_velocity: velocity:3822.16 dt: 0.02
update_velocity: velocity:3631.05 dt: 0.02
update_velocity: velocity:3449.50 dt: 0.02
update_velocity: velocity:3277.02 dt: 0.02
update_velocity: velocity:3113.17 dt: 0.01
update_velocity: velocity:2957.51 dt: 0.02
update_velocity: velocity:2809.64 dt: 0.02
update_velocity: velocity:2669.16 dt: 0.02
update_velocity: velocity:2535.70 dt: 0.02
update_velocity: velocity:2408.91 dt: 0.02
update_velocity: velocity:2288.47 dt: 0.01
update_velocity: velocity:2174.04 dt: 0.02
update_velocity: velocity:2065.34 dt: 0.02
update_velocity: velocity:1962.08 dt: 0.02
update_velocity: velocity:1863.97 dt: 0.02
update_velocity: velocity:1770.77 dt: 0.02
update_velocity: velocity:1682.23 dt: 0.02
update_velocity: velocity:1598.12 dt: 0.02
update_velocity: velocity:1518.22 dt: 0.02
update_velocity: velocity:1442.31 dt: 0.02
update_velocity: velocity:1370.19 dt: 0.02
update_velocity: velocity:1301.68 dt: 0.02
update_velocity: velocity:1236.60 dt: 0.01
update_velocity: velocity:1174.77 dt: 0.02
update_velocity: velocity:1116.03 dt: 0.02
update_velocity: velocity:1060.23 dt: 0.02
update_velocity: velocity:1007.22 dt: 0.02

This is what i get on Samsung S5

03-19 22:53:09.383 25442 25475 I python : update_velocity: velocity:7911.48 dt: 0.12
03-19 22:53:09.513 25442 25475 I python : update_velocity: velocity: 0.00 dt: 0.12
03-19 22:53:09.513 25442 25475 I python : update_velocity: velocity:7515.91 dt: 0.12
03-19 22:53:09.643 25442 25475 I python : update_velocity: velocity:7140.11 dt: 0.14
03-19 22:53:09.783 25442 25475 I python : update_velocity: velocity:6783.11 dt: 0.14
03-19 22:53:09.933 25442 25475 I python : update_velocity: velocity:6443.95 dt: 0.15
03-19 22:53:10.063 25442 25475 I python : update_velocity: velocity:6121.76 dt: 0.13
03-19 22:53:10.193 25442 25475 I python : update_velocity: velocity:5815.67 dt: 0.13
03-19 22:53:10.323 25442 25475 I python : update_velocity: velocity:5524.88 dt: 0.12
03-19 22:53:10.423 25442 25475 I python : update_velocity: velocity:5248.64 dt: 0.11
03-19 22:53:10.473 25442 25475 I python : update_velocity: velocity:4986.21 dt: 0.04
03-19 22:53:10.613 25442 25475 I python : update_velocity: velocity:4736.90 dt: 0.15
03-19 22:53:10.663 25442 25475 I python : update_velocity: velocity:4500.05 dt: 0.04
03-19 22:53:10.743 25442 25475 I python : update_velocity: velocity:4275.05 dt: 0.08
03-19 22:53:10.773 25442 25475 I python : update_velocity: velocity:4061.30 dt: 0.04
03-19 22:53:10.813 25442 25475 I python : update_velocity: velocity:3858.23 dt: 0.04
03-19 22:53:10.853 25442 25475 I python : update_velocity: velocity:3665.32 dt: 0.04
03-19 22:53:10.893 25442 25475 I python : update_velocity: velocity:3482.06 dt: 0.04
03-19 22:53:10.973 25442 25475 I python : update_velocity: velocity:3307.95 dt: 0.08
03-19 22:53:11.003 25442 25475 I python : update_velocity: velocity:3142.55 dt: 0.04
03-19 22:53:11.043 25442 25475 I python : update_velocity: velocity:2985.43 dt: 0.04
03-19 22:53:11.093 25442 25475 I python : update_velocity: velocity:2836.16 dt: 0.05
03-19 22:53:11.123 25442 25475 I python : update_velocity: velocity:2694.35 dt: 0.04
03-19 22:53:11.163 25442 25475 I python : update_velocity: velocity:2559.63 dt: 0.04
03-19 22:53:11.263 25442 25475 I python : update_velocity: velocity:2431.65 dt: 0.10
03-19 22:53:11.303 25442 25475 I python : update_velocity: velocity:2310.07 dt: 0.04
03-19 22:53:11.333 25442 25475 I python : update_velocity: velocity:2194.56 dt: 0.04
03-19 22:53:11.383 25442 25475 I python : update_velocity: velocity:2084.84 dt: 0.04
03-19 22:53:11.423 25442 25475 I python : update_velocity: velocity:1980.59 dt: 0.04
03-19 22:53:11.453 25442 25475 I python : update_velocity: velocity:1881.56 dt: 0.04
03-19 22:53:11.493 25442 25475 I python : update_velocity: velocity:1787.49 dt: 0.04
03-19 22:53:11.533 25442 25475 I python : update_velocity: velocity:1698.11 dt: 0.04
03-19 22:53:11.563 25442 25475 I python : update_velocity: velocity:1613.21 dt: 0.04
03-19 22:53:11.603 25442 25475 I python : update_velocity: velocity:1532.55 dt: 0.04
03-19 22:53:11.733 25442 25475 I python : update_velocity: velocity:1455.92 dt: 0.13
03-19 22:53:11.773 25442 25475 I python : update_velocity: velocity:1383.12 dt: 0.04
03-19 22:53:11.843 25442 25475 I python : update_velocity: velocity:1313.97 dt: 0.07
03-19 22:53:11.883 25442 25475 I python : update_velocity: velocity:1248.27 dt: 0.04
03-19 22:53:11.923 25442 25475 I python : update_velocity: velocity:1185.85 dt: 0.04
03-19 22:53:11.973 25442 25475 I python : update_velocity: velocity:1126.56 dt: 0.05
03-19 22:53:12.043 25442 25475 I python : update_velocity: velocity:1070.23 dt: 0.07
03-19 22:53:12.093 25442 25475 I python : update_velocity: velocity:1016.72 dt: 0.05
03-19 22:53:12.153 25442 25475 I python : update_velocity: velocity: 965.89 dt: 0.06
03-19 22:53:12.203 25442 25475 I python : update_velocity: velocity: 917.59 dt: 0.05
03-19 22:53:12.253 25442 25475 I python : update_velocity: velocity: 871.71 dt: 0.05
03-19 22:53:12.323 25442 25475 I python : update_velocity: velocity: 828.13 dt: 0.07
03-19 22:53:12.383 25442 25475 I python : update_velocity: velocity: 786.72 dt: 0.06
03-19 22:53:12.433 25442 25475 I python : update_velocity: velocity: 747.38 dt: 0.05
03-19 22:53:12.533 25442 25475 I python : update_velocity: velocity: 710.01 dt: 0.11
03-19 22:53:12.573 25442 25475 I python : update_velocity: velocity: 674.51 dt: 0.03
03-19 22:53:12.603 25442 25475 I python : update_velocity: velocity: 640.79 dt: 0.03
03-19 22:53:12.653 25442 25475 I python : update_velocity: velocity: 608.75 dt: 0.04
03-19 22:53:12.683 25442 25475 I python : update_velocity: velocity: 578.31 dt: 0.03
03-19 22:53:12.723 25442 25475 I python : update_velocity: velocity: 549.40 dt: 0.04
03-19 22:53:12.753 25442 25475 I python : update_velocity: velocity: 521.93 dt: 0.04
03-19 22:53:12.793 25442 25475 I python : update_velocity: velocity: 495.83 dt: 0.04
03-19 22:53:12.833 25442 25475 I python : update_velocity: velocity: 471.04 dt: 0.04
03-19 22:53:12.863 25442 25475 I python : update_velocity: velocity: 447.49 dt: 0.04
03-19 22:53:12.903 25442 25475 I python : update_velocity: velocity: 425.11 dt: 0.04
03-19 22:53:12.933 25442 25475 I python : update_velocity: velocity: 403.86 dt: 0.04
03-19 22:53:12.973 25442 25475 I python : update_velocity: velocity: 383.66 dt: 0.04
03-19 22:53:13.003 25442 25475 I python : update_velocity: velocity: 364.48 dt: 0.04
03-19 22:53:13.043 25442 25475 I python : update_velocity: velocity: 346.26 dt: 0.04
03-19 22:53:13.083 25442 25475 I python : update_velocity: velocity: 328.94 dt: 0.04
03-19 22:53:13.123 25442 25475 I python : update_velocity: velocity: 312.50 dt: 0.04
03-19 22:53:13.153 25442 25475 I python : update_velocity: velocity: 296.87 dt: 0.04

### CODE REVIEW

File: kivy/effects/kinetic.py [hunk 1]
```diff
@@ -177,6 +189,6 @@ def update_velocity(self, dt):
             self.velocity = 0
             return
 
-        self.velocity -= self.velocity * self.friction
+        self.velocity -= self.velocity * self.friction * dt / self.std_dt
```
Comment (matham) [inline kivy/effects/kinetic.py:192]: How is this different than simply setting `friction` to `0.05/0.017`?

Comment (td1803) [reply kivy/effects/kinetic.py:192]: The difference is the consideration of the parameter dt. 
This value describes the elapsed time and is not constant. 
See the values for dt above.

Therefore, friction must always be applied over time and not as a constant. The value self.std_dt here only ensures that the previous value for friction can still be used.

Comment (tshirtman) [reply kivy/effects/kinetic.py:192]: seems sensible to me, correcting for the actual dt vs expected dt should give a more consistent experience accross devices.

**Review by tshirtman — APPROVED** [review]

File: kivy/effects/kinetic.py (hunk 1 above)
Comment (matham) [reply kivy/effects/kinetic.py:192]: Ok. What confused me is that when searching I only saw `friction` used with `dt`. So it wasn't clear to me why you couldn't replace `self.friction * dt / self.std_dt` with `self.friction * dt` but simply adjust `friction` to account for the variance on each device.

But I guess you're saying that conceptually they have different meanings so conceptually we should be able to set both `friction` and `std_dt` independently, even if in practice they are only ever used together as one factor? If I understood this correctly, it should be ok.

**Review by matham — APPROVED** [review]

**Comment by tshirtman** [comment]
I agree that making the value configurable is possibly unneccessary, but it doesn't hurt either, what's important imho is to be able to compare the actual dt with a reference value (configurable or not), and adjust the friction based on that.
The only alternative is to have fixed timesteps, but it's a hard promise to make as a framework, as we can't prevent users from slowing down our event loop, and even an hello world app can have a different fps depending on where it's running.
Also, providing this doesn't prevent a cautious developper from actually fixing their timestep, and having a better result than what we can provide with this change, and less cautious developpers will benefit from the improvement anyway.

**Pull request merged**